client/chat/channel.tsx

```diff
@@ -463,109 +461,66 @@ class Channel extends React.Component<ChannelProps> {
   }
 }
 
-const mapStateToProps = (state: RootState) => {
-  return {
-    user: state.auth.user,
-    chat: state.chat,
-  }
-}
+export default function ChatChannelView() {
+  const [, params = {}] = useRoute('/chat/:channel')
+  const channelName = params ? decodeURIComponent(params.channel).toLowerCase() : null
+  const dispatch = useAppDispatch()
+  const channel = useAppSelector(s => s.chat.byName.get(channelName))
 
-interface ChatChannelViewProps {
-  user: SelfUserRecord
-  chat: ReturnType<typeof ChatState>
-  dispatch: DispatchFunction<ReduxAction>
-  params: { channel: string }
-}
+  const prevChannelName = usePrevious(channelName)
+  const prevChannel = usePrevious(channel)
 
-function isLeavingChannel(oldProps: ChatChannelViewProps, newProps: ChatChannelViewProps) {
-  return (
-    oldProps.params.channel.toLowerCase() === newProps.params.channel.toLowerCase() &&
-    oldProps.chat.byName.has(oldProps.params.channel.toLowerCase()) &&
-    !newProps.chat.byName.has(oldProps.params.channel.toLowerCase())
-  )
-}
-
-class ChatChannelView extends React.Component<ChatChannelViewProps> {
-  componentDidMount() {
-    const routeChannel = decodeURIComponent(this.props.params.channel)
-    if (this._isInChannel()) {
-      this.props.dispatch(retrieveUserList(routeChannel))
-      this.props.dispatch(retrieveInitialMessageHistory(routeChannel))
-      this.props.dispatch(activateChannel(routeChannel) as any)
-    } else {
-      this.props.dispatch(joinChannel(routeChannel))
-    }
-  }
-
-  componentDidUpdate(prevProps: ChatChannelViewProps) {
-    if (isLeavingChannel(prevProps, this.props)) {
+  // TODO(2Pac): Pull this out into some kind of "isLeaving" hook and share with whispers/lobby?
+  useEffect(() => {
+    if (prevChannelName === channelName && prevChannel && !channel) {
       push('/')
-      return
     }
+  })
 
-    const prevChannel = decodeURIComponent(prevProps.params.channel)
-    const routeChannel = decodeURIComponent(this.props.params.channel)
-    if (this._isInChannel()) {
-      this.props.dispatch(retrieveUserList(routeChannel))
-      this.props.dispatch(retrieveInitialMessageHistory(routeChannel))
-      this.props.dispatch(activateChannel(routeChannel) as any)
-    } else if (
-      !prevProps.chat.byName.has(routeChannel) &&
-      prevChannel.toLowerCase() !== routeChannel.toLowerCase()
-    ) {
+  // TODO(2Pac): Make this less spammy?
+  useEffect(() => {
+    dispatch(activateChannel(channelName) as any)
+  }, [channel])
+
+  useEffect(() => {
+    if (channel) {
```

Comment:
There's a lint rule that checks these but imo it has a lot of false positives, but generally:

Don't use values from outside `useEffect` that you haven't declared as part of the dependencies. I can see why you'd think this would be safe, but it actually isn't. If you navigate to a channel you haven't yet joined, this will run once and join the channel. Then it will never run again, because joining the channel doesn't change the channel name. So the following code that we expect to run (retrieving the user list, retrieving the message history) will never occur

I would probably add `channel` to the deps, add additional checks inside the effect to ensure these only run on actual channel changes. Probably you should combine all these disparate effects back into a single one that checks all these things together, more similarly to the `componentDidUpdate`. I understand the appeal of splitting things up, but these things likely are actually dependent on each other and should be together.

Reply:
After thinking a bit more about this, I realized my comment above is wrong on when should we be activating the channel. It totally makes sense to activate the channel only after we retrieve the initial message history as we've been doing. However, the reason I couldn't make that work is because we were depending on _any_ change to the `channel` record in the store, which makes things more complicated than they should be.

For example, when joining the channel, the `channel` record gets updated four distinct times and each time we execute code to retrieve message history, user list, and activate the channel. Doing that was mostly fine because those functions have failsafes on when to actually preform the action (except the channel activation one, which does it each time, thus making the action history a bit spammy). But since now with hooks we were also deactivating channel after each render, depending on _any_ change to the `channel` record doesn't seem to make much more sense anymore.

I feel like we could easily deal with this by simply calculating values that we're interested in outside the effect function, and then depend on only those values, instead of the whole `channel` record. Something like:
```js
const isInChannel = !!channel
const isLeavingChannel = !isInChannel && prevChannel && prevChannelName === channelName

useEffect(() => {
  if (isLeavingChannel) {
    push('/')
  }
}, [isLeavingChannel])

useEffect(() => {
  if (isInChannel) {
    dispatch(retrieveUserList(channelName))
    dispatch(retrieveInitialMessageHistory(channelName))
    dispatch(activateChannel(channelName) as any)
  } else if (!isLeavingChannel) {
    if (MULTI_CHANNEL) {
      dispatch(joinChannel(channelName))
    } else {
      push('/')
    }
  }

  return () => dispatch(deactivateChannel(channelName) as any)
}, [isInChannel, isLeavingChannel, channelName])
```
I tested this and it seems to be doing exactly what we need, but I'm not sure if I'm missing something.

Reply:
Yeah, that looks much more correct to me 👍 

In general this is also why it's better to grab granular things out of the state rather than the entire record (also helps optimize re-renders from `useAppSelector`)